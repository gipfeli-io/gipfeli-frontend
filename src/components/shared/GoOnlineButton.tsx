import { Fab } from '@mui/material'
import React from 'react'
import WifiIcon from '@mui/icons-material/Wifi'
import useConnectionStatus from '../../hooks/use-connection-status'
import useGoOnline from '../../hooks/use-go-online'
import useTheme from '../../hooks/use-theme'

const GoOnlineButton = () => {
  const { showGoOnlineButton } = useConnectionStatus()
  const activateOnlineMode = useGoOnline()
  const { activeTheme: { theme } } = useTheme()

  const goOnline = async (): Promise<void> => {
    (await activateOnlineMode)()
  }

  return (
    showGoOnlineButton
      ? (
        <Fab
          variant="extended"
          onClick={goOnline}
          color="primary"
          sx={{ position: 'fixed', bottom: theme.spacing(3), right: theme.spacing(3) }}
        >
          <WifiIcon sx={{ mr: 1 }}/>
          Go online
        </Fab>)

      : null
  )
}

export default GoOnlineButton
