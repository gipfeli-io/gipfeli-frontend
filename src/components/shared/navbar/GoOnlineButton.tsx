import { Button } from '@mui/material'
import React from 'react'
import useConnectionStatus from '../../../hooks/use-connection-status'
import WifiIcon from '@mui/icons-material/Wifi'
import useGoOnline from '../../../hooks/use-go-online'

const GoOnlineButton = () => {
  const { showGoOnlineButton } = useConnectionStatus()
  const activateOnlineMode = useGoOnline()
  const goOnline = async (): Promise<void> => {
    (await activateOnlineMode)()
  }
  return (
    showGoOnlineButton
      ? <Button sx={{ mr: 3 }} onClick={goOnline} variant="outlined" color="inherit" startIcon={<WifiIcon/>}>Go online</Button>
      : null
  )
}

export default GoOnlineButton
