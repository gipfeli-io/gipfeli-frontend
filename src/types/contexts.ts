import { Notification } from './notifications'
import { AppTheme } from './theme'
import { handleSave } from './handle-save'
import { CurrentUpload, GpxFileUpload, ImageUpload } from './media'
import { ConnectionStatus } from '../enums/connection-status'
import { ErrorInfo } from 'react'

export interface TourListContextProperties {
  deleteEvent: (id: string) => void
}

export interface ListContextProperties {
  deleteEvent: (id: string) => void
}

export interface AuthenticationContextType {
  email: string | undefined
  isAdmin: boolean
  token: string | undefined
  signIn: (email: string, password: string, callback: () => void) => Promise<void>
  signOut: (callback: () => void) => void
}

export interface NotificationContextType {
  notification: Notification | undefined
  triggerErrorNotification: (message: string) => void
  triggerSuccessNotification: (message: string) => void
  triggerOfflineNotification: () => void
  triggerSyncFailedNotification: (message: string) => void
  resetNotification: () => void
}

export interface ConnectionStatusContextType {
  isOffline: () => boolean
  updateConnectionStatus: (connectionStatus: ConnectionStatus) => void
  updateGoOnlineButtonVisibility: (isVisible: boolean) => void
  showGoOnlineButton: boolean
  isOnlineInfoBannerVisible: boolean
  updateOnlineInfoBannerVisibility: (isVisible: boolean) => void
  resetOnlineInfoBanner: () => void
}

export interface ThemeContextType {
  activeTheme: AppTheme
  toggleTheme: () => void
}

export interface ImageUploadContextType {
  files: ImageUpload[]
  save: handleSave<File[]>
  remove: (id: string) => void
  currentUploads: CurrentUpload[]
}

export interface GpxFileUploadContextType {
  file: GpxFileUpload
  save: handleSave<File>
  remove: (id: string) => void
  currentUpload?: CurrentUpload
}

export interface ErrorBoundaryContextType{
  triggerError: (error: Error, errorInfo?: ErrorInfo) => void
}
