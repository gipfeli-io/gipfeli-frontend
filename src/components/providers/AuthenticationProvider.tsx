import React, { PropsWithChildren, useEffect, useMemo, useState } from 'react'
import AuthenticationContext from '../../contexts/authentication-context'
import AuthService from '../../services/auth/auth-service'
import LocalStorageService from '../../services/local-storage-service'
import { LocalStorageKey } from '../../enums/local-storage-key'
import jwtDecode from 'jwt-decode'
import Loader from '../shared/Loader'
import { JwtToken } from '../../types/jwt-token'
import useApiError from '../../hooks/use-api-error'
import useNotifications from '../../hooks/use-notifications'
import { AuthenticationContextType } from '../../types/contexts'
import useInterval from '../../hooks/use-interval'
import tokenNeedsRefresh from '../../utils/token-needs-refresh'
import { AccessToken, AuthObject } from '../../types/auth'
import { useNavigate } from 'react-router'
import { UserRole } from '../../enums/user-role'

const AuthenticationProvider = ({ children }: PropsWithChildren<any>) => {
  const [email, setEmail] = useState<string | undefined>(undefined)
  const [userRole, setUserRole] = useState<UserRole | undefined>(undefined)
  const [userId, setUserId] = useState<string | undefined>(undefined)
  const [accessToken, setAccessToken] = useState<string | undefined>(undefined)
  const [refreshToken, setRefreshToken] = useState<string | undefined>(undefined)
  const [loading, setLoading] = useState<boolean>(true)
  const authService: AuthService = new AuthService()
  const localStorageService: LocalStorageService = new LocalStorageService()
  const throwError = useApiError()
  const { triggerErrorNotification } = useNotifications()
  const navigate = useNavigate()

  const setTokensInLocalStorageAndState = (tokens: AuthObject) => {
    const { accessToken, refreshToken } = tokens
    localStorageService.addItem(LocalStorageKey.AccessToken, accessToken)
    localStorageService.addItem(LocalStorageKey.RefreshToken, refreshToken)
    setAccessToken(accessToken)
    setRefreshToken(refreshToken)
    const { email, role, sub } = jwtDecode<AccessToken>(accessToken)
    setEmail(email)
    setUserRole(role)
    setUserId(sub)
  }

  const unsetTokensInLocalStorageAndState = () => {
    localStorageService.removeItem(LocalStorageKey.AccessToken)
    localStorageService.removeItem(LocalStorageKey.RefreshToken)
    setAccessToken(undefined)
    setRefreshToken(undefined)
    setEmail(undefined)
    setUserRole(undefined)
    setUserId(undefined)
  }

  const isAdmin = useMemo(() => {
    if (userRole === undefined) {
      return false
    }

    return userRole === UserRole.ADMINISTRATOR
  }, [userRole])

  const signIn = async (email: string, password: string, callback: () => void) => {
    const data = await authService.login(
      email,
      password
    )

    if (data.success) {
      setTokensInLocalStorageAndState(data.content!)
      callback()
    } else if (data.statusCode === 404) {
      triggerErrorNotification('Combination of password and user does not exist.')
    } else {
      throwError(data)
      callback()
    }
  }

  const signOut = async (callback: () => void) => {
    // Todo: do we need a signout request at all?
    unsetTokensInLocalStorageAndState()
    callback()
  }

  const forceLogOut = () => {
    unsetTokensInLocalStorageAndState()
    triggerErrorNotification('You have been logged out due to inactivity.')
    navigate('/login')
  }

  /**
   * Create a request to refresh our tokens - or force logout.
   * @param refreshToken
   */
  const fetchNewTokens = async (refreshToken: string) => {
    const data = await authService.refreshTokens(refreshToken)

    if (data.success) {
      setTokensInLocalStorageAndState(data.content!)
    } else {
      forceLogOut()
    }
  }

  /**
   * Checks whether a token is still valid and handles its refresh or force logout.
   */
  const checkAndRefreshToken = () => {
    // We only check for the token if we actually have a token
    if (!accessToken) {
      return
    }
    const decodedToken: JwtToken = jwtDecode(accessToken)

    if (tokenNeedsRefresh(decodedToken)) {
      if (refreshToken) {
        const refresh = async () => {
          await fetchNewTokens(refreshToken)
        }

        refresh()
      } else {
        forceLogOut()
      }
    }
  }

  /**
   * Handle a new page load and check whether we still have a session or not.
   */
  useEffect(() => {
    const accessToken = localStorageService.getItem(LocalStorageKey.AccessToken)
    const refreshToken = localStorageService.getItem(LocalStorageKey.RefreshToken)

    if (accessToken && refreshToken) {
      const decodedAccessToken: JwtToken = jwtDecode<AccessToken>(accessToken)
      if (tokenNeedsRefresh(decodedAccessToken)) {
        const refresh = async () => {
          await fetchNewTokens(refreshToken)
          setLoading(false)
        }

        refresh()
      } else {
        setTokensInLocalStorageAndState({ accessToken, refreshToken })
        setLoading(false)
      }
    } else {
      setLoading(false)
    }
  }, [])

  /**
   * Register our refresh token interval.
   */
  useInterval(checkAndRefreshToken, 2000)

  const value: AuthenticationContextType = { userId, email, isAdmin, token: accessToken, signIn, signOut }

  if (loading) {
    return <Loader/>
  }

  return <AuthenticationContext.Provider value={value}>{children}</AuthenticationContext.Provider>
}

export default AuthenticationProvider
