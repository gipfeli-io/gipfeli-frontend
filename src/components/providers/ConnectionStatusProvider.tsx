import React, { PropsWithChildren, useState } from 'react'
import ConnectionStatusContext from '../../contexts/connection-status-context'
import { ConnectionStatus } from '../../enums/connection-status'
import { ConnectionStatusContextType } from '../../types/contexts'
import { LocalStorageKey } from '../../enums/local-storage-key'
import LocalStorageService from '../../services/local-storage-service'
import { useNavigate } from 'react-router'

export const ConnectionStatusProvider = ({ children }: PropsWithChildren<any>) => {
  const localStorageService = new LocalStorageService()
  const navigate = useNavigate()
  const requestUrl: string = process.env.REACT_APP_PUBLIC_BACKEND_API || 'http://localhost:3000'
  const pollingDelay: string | number = process.env.ONLINE_POLLING_DELAY || 20000
  // eslint-disable-next-line no-undef
  let intervalId: NodeJS.Timer

  const getInitialOnlineInfoBannerVisibility = () => {
    const savedOnlineInfoBannerVisibility = localStorageService.getItem(LocalStorageKey.IsOnlineBannerSeen)
    if (!savedOnlineInfoBannerVisibility) {
      return false
    }
    return JSON.parse(savedOnlineInfoBannerVisibility) === true
  }

  const [isOnlineInfoBannerVisible, setIsOnlineInfoBannerVisible] = useState<boolean>(getInitialOnlineInfoBannerVisibility)

  const updateOnlineInfoBannerVisibility = (isVisible: boolean) => {
    localStorageService.addItem(LocalStorageKey.IsOnlineBannerSeen, String(!isVisible))
    setIsOnlineInfoBannerVisible(isVisible)
  }

  const resetOnlineInfoBanner = () => {
    localStorageService.removeItem(LocalStorageKey.IsOnlineBannerSeen)
    setIsOnlineInfoBannerVisible(false)
  }

  const getInitialGoOnlineButtonState = () => {
    const savedGoOnlineButtonState = localStorageService.getItem(LocalStorageKey.IsGoOnlineButtonVisible)
    if (!savedGoOnlineButtonState) {
      return false
    }

    return JSON.parse(savedGoOnlineButtonState) === true
  }

  const [showGoOnlineButton, setShowGoOnlineButton] = useState<boolean>(getInitialGoOnlineButtonState)

  const updateGoOnlineButtonVisibility = (isVisible: boolean) => {
    localStorageService.addItem(LocalStorageKey.IsGoOnlineButtonVisible, String(isVisible))
    setShowGoOnlineButton(isVisible)
  }

  const checkIfApplicationIsOnline = async () : Promise<void> => {
    const request = await fetch(requestUrl + '/heartbeat')
    if (request.status === 200) {
      clearInterval(intervalId)
      updateGoOnlineButtonVisibility(true)
      updateOnlineInfoBannerVisibility(true)
    }
  }

  const initPolling = () => {
    intervalId = setInterval(checkIfApplicationIsOnline, +pollingDelay)
  }

  const getInitialConnectionStatus = () => {
    const savedConnectionStatusChoice = localStorageService.getItem(LocalStorageKey.ConnectionStatus)
    if (savedConnectionStatusChoice) {
      const status = savedConnectionStatusChoice === ConnectionStatus.ONLINE ? ConnectionStatus.ONLINE : ConnectionStatus.OFFLINE
      if (status === ConnectionStatus.OFFLINE) {
        initPolling()
      }
      return status
    }

    return ConnectionStatus.ONLINE
  }

  const [connectionStatus, setConnectionStatus] = useState<ConnectionStatus>(getInitialConnectionStatus)

  const updateConnectionStatus = (status: ConnectionStatus) => {
    localStorageService.addItem(LocalStorageKey.ConnectionStatus, status)
    setConnectionStatus(status)
    if (status === ConnectionStatus.OFFLINE) {
      initPolling()
    } else {
      updateGoOnlineButtonVisibility(false)
    }
    navigate('/tours', { replace: true })
  }

  const isOffline = () => connectionStatus === ConnectionStatus.OFFLINE

  const value: ConnectionStatusContextType = {
    isOffline,
    updateConnectionStatus,
    updateGoOnlineButtonVisibility,
    showGoOnlineButton,
    isOnlineInfoBannerVisible,
    updateOnlineInfoBannerVisibility,
    resetOnlineInfoBanner
  }

  return (
    <ConnectionStatusContext.Provider value={value}>
      {children}
    </ConnectionStatusContext.Provider>
  )
}
