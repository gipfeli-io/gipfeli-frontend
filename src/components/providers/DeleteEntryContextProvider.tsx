import React, { PropsWithChildren, useState } from 'react'
import { DeleteEntryContextType } from '../../types/contexts'
import DeleteEntryContext from '../../contexts/delete-entry-context'

type DeleteEntryContextProviderProps = {
  id?: string
}

const DeleteEntryContextProvider = ({ children, id }: PropsWithChildren<DeleteEntryContextProviderProps>) => {
  const [showDeleteModal, setShowDeleteModal] = useState(false)
  const [deleteId, setDeleteId] = useState<string | undefined>(undefined)

  const value: DeleteEntryContextType = {
    showDeleteModal,
    toggleModal: () => setShowDeleteModal(prevState => !prevState),
    deleteId: id ?? deleteId,
    setDeleteId: (id) => setDeleteId(id)
  }

  return (
    <DeleteEntryContext.Provider value={value}>
      {children}
    </DeleteEntryContext.Provider>
  )
}

export default DeleteEntryContextProvider
