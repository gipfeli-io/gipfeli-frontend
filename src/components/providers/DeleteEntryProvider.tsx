import React, { PropsWithChildren, useState } from 'react'
import { DeleteEntryContextType } from '../../types/contexts'
import DeleteEntryContext from '../../contexts/delete-entry-context'

type DeleteEntryProviderProps = {
  id?: string
}

const DeleteEntryProvider = ({ children, id }: PropsWithChildren<DeleteEntryProviderProps>) => {
  const [showDeleteModal, setShowDeleteModal] = useState(false)
  const [deleteId, setDeleteId] = useState<string | undefined>(undefined)

  const value: DeleteEntryContextType = {
    showDeleteModal,
    toggleModal: () => setShowDeleteModal(prevState => !prevState),
    deleteId: id ?? deleteId,
    setDeleteId: (id) => setDeleteId(id)
  }

  return (
    <DeleteEntryContext.Provider value={value}>
      {children}
    </DeleteEntryContext.Provider>
  )
}

export default DeleteEntryProvider
