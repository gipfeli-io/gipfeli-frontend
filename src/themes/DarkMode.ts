import { createTheme } from '@mui/material/styles'
import { LocalStorageKey } from '../enums/LocalStorageKey'
import { AppTheme } from '../types/theme'

const DarkMode: AppTheme = {
  name: LocalStorageKey.DarkMode,
  theme: createTheme({
    palette: {
      mode: 'dark'
    }
  })
}

export default DarkMode
