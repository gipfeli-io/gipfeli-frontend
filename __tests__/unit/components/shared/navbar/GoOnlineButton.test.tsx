import '@testing-library/jest-dom'
// @ts-ignore
import React from 'react'
import 'reflect-metadata'
// @ts-ignore
import renderer from 'react-test-renderer'
import GoOnlineButton from '../../../../../src/components/shared/GoOnlineButton'
import { ConnectionStatusProvider } from '../../../../../src/components/providers/ConnectionStatusProvider'
import { ThemeContextType } from '../../../../../src/types/contexts'
import DarkMode from '../../../../../src/themes/dark-mode'

const mockConnectionStatusContext: any = {
  showGoOnlineButton: false
}

const triggerSyncFailedNotification: jest.Mock = jest.fn()
const mockNotificationContext: any = {
  triggerSyncFailedNotification
}

const mockAuthContext: any = {
  token: undefined
}

const errorBoundaryContext: any = {
  triggerError: jest.fn()
}

const mockUseLocationValue = {
  pathname: '/localhost:3000/tours',
  search: '',
  hash: '',
  state: null
}

const mockThemeContext: ThemeContextType = {
  activeTheme: DarkMode,
  toggleTheme: jest.fn()
}

jest.mock('../../../../../src/hooks/use-connection-status', () => jest.fn().mockImplementation(() => mockConnectionStatusContext))
jest.mock('../../../../../src/hooks/use-notifications', () => jest.fn().mockImplementation(() => mockNotificationContext))
jest.mock('../../../../../src/hooks/use-theme', () => jest.fn().mockImplementation(() => mockThemeContext))
jest.mock('../../../../../src/hooks/use-auth', () => jest.fn().mockImplementation(() => mockAuthContext))
jest.mock('../../../../../src/hooks/use-error-handling', () => jest.fn().mockImplementation(() => errorBoundaryContext))

jest.mock('react-router', () => ({
  ...jest.requireActual('react-router') as any,
  useNavigate: () => jest.fn(),
  useLocation: () => jest.fn().mockImplementation(() => mockUseLocationValue),
  useHref: () => jest.fn()
}))

describe('GoOnlineButton', () => {
  it('will show button if user offline with connection', () => {
    mockConnectionStatusContext.showGoOnlineButton = true
    const tree = renderer
      .create(<ConnectionStatusProvider><GoOnlineButton/></ConnectionStatusProvider>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('will not show button if user is online', () => {
    mockConnectionStatusContext.showGoOnlineButton = false
    const tree = renderer
      .create(<ConnectionStatusProvider><GoOnlineButton/></ConnectionStatusProvider>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
