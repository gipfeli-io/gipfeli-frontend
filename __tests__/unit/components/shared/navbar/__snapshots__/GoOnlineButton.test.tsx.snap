// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoOnlineButton will not show button if user is online 1`] = `null`;

exports[`GoOnlineButton will show button if user offline with connection 1`] = `
<button
  className="MuiButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorInherit MuiButtonBase-root  css-1vsri9o-MuiButtonBase-root-MuiButton-root"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="WifiIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m1 9 2 2c4.97-4.97 13.03-4.97 18 0l2-2C16.93 2.93 7.08 2.93 1 9zm8 8 3 3 3-3c-1.65-1.66-4.34-1.66-6 0zm-4-4 2 2c2.76-2.76 7.24-2.76 10 0l2-2C15.14 9.14 8.87 9.14 5 13z"
      />
    </svg>
  </span>
  Go online
</button>
`;
